Csocket.cc

```diff
@@ -2691,39 +2746,56 @@ CS_STRING Csock::GetPeerPubKey() const
 {
 	CS_STRING sKey;
 
-	SSL_SESSION * pSession = GetSSLSession();
+	X509 * pCert = GetX509();
 
-	if( pSession && pSession->peer )
+	if( pCert )
 	{
-		EVP_PKEY * pKey = X509_get_pubkey( pSession->peer );
+		EVP_PKEY * pKey = X509_get_pubkey( pCert );
 		if( pKey )
 		{
-			char *hxKey = NULL;
-			switch( pKey->type )
+			const BIGNUM * pPubKey = NULL;
+#ifdef HAVE_OPAQUE_EVP_PKEY
+			int iType = EVP_PKEY_base_id( pKey );
+#else
+			int iType = pKey->type;
+#endif /* HAVE_OPAQUE_EVP_PKEY */
+			switch( iType )
 			{
+#ifndef OPENSSL_NO_RSA
 			case EVP_PKEY_RSA:
-			{
-				hxKey = BN_bn2hex( pKey->pkey.rsa->n );
+# ifdef HAVE_OPAQUE_RSA
```

Comment:
```c++
#ifndef ...
# define EVP_PKEY_get0_RSA(pKey) ((pKey)->pkey.rsa)
#endif
```

Something similar for `RSA_get0_key` but as inline function. `EVP_PKEY_get0_RSA` also can be defined as inline function instead.

Was there any OpenSSL version released which has one opaque struct, but not the other? If not, I don't think it's necessary to support such case, which would simplify code.